accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
++       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();
+    }
+
+    public abstract class FieldInfo
+    {
++       [System.CLSCompliantAttribute(false)]
+        public virtual void GetValueDirect(TypedReference obj, TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
+        public virtual void SetValueDirect(TypedReference obj, TypedReference value);
+
+        // Possible for performance in System.Text.Json:
++       public virtual Func<object, TValue> CreateGetterDelegate<TValue>();
++       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();
+    }
+}
+```
+
+## Examples
+### Fixed-length arguments
+```cs
+MethodInfo method = ... // Some method to call
+ArgumentValuesFixed values = new(4); // 4 parameters
+InvokeContext context = new InvokeContext(ref values);
+context.SetArgument(0, new MyClass());
+context.SetArgument(1, null);
+context.SetArgument(2, 42);
+context.SetArgument(3, "Hello");
+
+// Can inspect before or after invoke:
+object o0 = context.GetArgument(0);
+object o1 = context.GetArgument(1);
+object o2 = context.GetArgument(2);
+object o3 = context.GetArgument(3);
+
+context.InvokeDirect(method);
+int ret = (int)context.GetReturn();
+```
+
+### Fixed-length object arguments (faster)
+```cs
+ArgumentValuesFixed args = new(new MyClass(), null, 42, "Hello");
+InvokeContext context = new InvokeContext(ref args);
+context.InvokeDirect(method);
+```
+
+### Variable-length object arguments
+Unsafe and slightly slower than fixed-length plus requires `using` or `try\finally\Dispose()`.
+```cs
+unsafe
+{
+    ArgumentValue* args = stackalloc ArgumentValue[4];
+    using (InvokeContext context = new InvokeContext(ref args))
+    {
+        context.SetArgument(0, new MyClass());
+        context.SetArgument(1, null);
+        context.SetArgument(2, 42);
+        context.SetArgument(3, "Hello");
+        context.InvokeDirect(method);
+    }
+}
+```
+
+### Avoiding boxing
+Value types can be references to avoid boxing.
+
+```cs
+int i = 42;
+int ret = 0;
+ArgumentValuesFixed args = new(4);
+InvokeContext context = new InvokeContext(ref args);
+context.SetArgument(0, new MyClass());
+context.SetArgument(1, null);
+context.SetArgument<int>(2, ref i); // No boxing (argument not required to be byref)
+context.SetArgument(3, "Hello");
+context.SetReturn<int>(ref ret); // No boxing; 'ret' variable updated automatically
+context.InvokeDirect(method);
+```
+
+### Pass a `Span<T>` to a method
+```cs
+Span<int> span = new int[] { 42, 43 };
+ArgumentValuesFixed args = new(1);
+
+unsafe
+{
+    InvokeContext context = new InvokeContext(ref args);
+#pragma warning disable CS8500    
+    void* ptr = (void*)new IntPtr(&span);
+#pragma warning restore CS8500    
+    // Ideally we can use __makeref(span) instead of the above.
+
+    context.SetArgument(0, ptr, typeof(Span<int>));
+    context.InvokeDirect(method);
+}
+```
+# Design ext
```

Comment:
"extension"?

Reply:
Will change to "addendum"; just placing less-important details and findings in an area that isn't up-front and cluttering.